{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "71f09632",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Chase\n",
    "# Initial list of RNA proteins.\n",
    "RNAList = {'UUU':'F', 'CUU':'L', 'AUU':'I', 'GUU':'V',\n",
    "'UUC':'F', 'CUC':'L', 'AUC':'I', 'GUC':'V',\n",
    "'UUA':'L', 'CUA':'L', 'AUA':'I', 'GUA':'V',\n",
    "'UUG':'L', 'CUG':'L', 'AUG':'M', 'GUG':'V',\n",
    "'UCU':'S', 'CCU':'P', 'ACU':'T', 'GCU':'A',\n",
    "'UCC':'S', 'CCC':'P', 'ACC':'T', 'GCC':'A',\n",
    "'UCA':'S', 'CCA':'P', 'ACA':'T', 'GCA':'A',\n",
    "'UCG':'S', 'CCG':'P', 'ACG':'T', 'GCG':'A',\n",
    "'UAU':'Y', 'CAU':'H', 'AAU':'N', 'GAU':'D',\n",
    "'UAC':'Y', 'CAC':'H', 'AAC':'N', 'GAC':'D',\n",
    "'UAA':' ', 'CAA':'Q', 'AAA':'K', 'GAA':'E',\n",
    "'UAG':' ', 'CAG':'Q', 'AAG':'K', 'GAG':'E',\n",
    "'UGU':'C', 'CGU':'R', 'AGU':'S', 'GGU':'G',\n",
    "'UGC':'C', 'CGC':'R', 'AGC':'S', 'GGC':'G',\n",
    "'UGA':' ', 'CGA':'R', 'AGA':'R', 'GGA':'G',\n",
    "'UGG':'W', 'CGG':'R', 'AGG':'R', 'GGG':'G',\n",
    "}\n",
    "\n",
    "#This function is responsible for providing the reverse compliment of a DNA sequence.\n",
    "def reverse_compliment(dna):\n",
    "    reverse = ''\n",
    "    for i in range(0,len(dna)):\n",
    "        if dna[i] == 'A':\n",
    "            reverse += 'T'\n",
    "        elif dna[i] == 'T':\n",
    "            reverse += 'A'\n",
    "        elif dna[i] == 'C':\n",
    "            reverse += 'G'\n",
    "        elif dna[i] == 'G':\n",
    "            reverse += 'C'\n",
    "    return reverse[::-1]       \n",
    "\n",
    "\n",
    "#This function returns the corresponding codon for an RNA protein sequence.\n",
    "def rna2codon(rna):\n",
    "    codon = ''\n",
    "    for i in range(0, len(rna), 3):\n",
    "        codon += RNAList[rna[i:i+3]]\n",
    "    return codon\n",
    "\n",
    "\n",
    "#This function counts the number of protein combinations that can produce a given codon. \n",
    "def source_rna(protein):\n",
    "    values = RNAList.values()\n",
    "    values_list = list(values)\n",
    "    count1 = 0\n",
    "    count2 = 1\n",
    "    proteinStop = protein + ' '\n",
    "    for x in range (0,len(proteinStop)):\n",
    "        for i in range (0,len(RNAList)):\n",
    "            if values_list[i] == proteinStop[x]:\n",
    "                count1 += 1\n",
    "        count2 *= count1\n",
    "        count1 = 0\n",
    "    return count2\n",
    "\n",
    "#Luis\n",
    "\n",
    "def locate_substring(dna_snippet, dna):\n",
    "    start = 0\n",
    "    end = len(dna)\n",
    "    x = 0\n",
    "    postion = []\n",
    "    for start in range (0, len(dna)):\n",
    "        start = x + 1\n",
    "        x = dna.find(dna_snippet, start, end)\n",
    "        if x > -1:\n",
    "            postion.append(x)\n",
    "        if x <= 0:\n",
    "            break\n",
    "    return(postion)\n",
    "\n",
    "def mendels_law(hom, het, rec):\n",
    "    g_total = (hom + het + rec)\n",
    "    n_total = (g_total*(g_total-1))*2\n",
    "    rec_total = (rec*(rec-1))*2\n",
    "    het_total = (het*(het-1))/2\n",
    "    mixed_total = het*rec*2\n",
    "    r_total = rec_total + het_total+ mixed_total\n",
    "    P_dom = 1 - (r_total / n_total)\n",
    "    return(P_dom)\n",
    "\n",
    "\n",
    "def locate_substring(dna_snippet, dna):\n",
    "    start = 0\n",
    "    end = len(dna)\n",
    "    x = 0\n",
    "    postion = []\n",
    "    for start in range (0, len(dna)):\n",
    "        start = x + 1\n",
    "        x = dna.find(dna_snippet, start, end)\n",
    "        if x > -1:\n",
    "            postion.append(x)\n",
    "        if x <= 0:\n",
    "            break\n",
    "    return(postion)\n",
    "\n",
    "#This is a new function unique to this one.\n",
    "def intron_cut(dna, intron_list):\n",
    "    i = 0\n",
    "    intron=[]\n",
    "    OrgDna=dna\n",
    "    L=0\n",
    "    # return the postion of the introns\n",
    "    for i in range(0,len(intron_list)):\n",
    "        dna_snippets = intron_list[i]\n",
    "        j = locate_substring(dna_snippets, dna)[0]\n",
    "        intron.append(j)\n",
    "        if i == len(intron_list):\n",
    "            break\n",
    "    for i in range(0,len(intron)):\n",
    "        One_slice = slice(intron[i]-L,intron[i]+len(intron_list[i])-L)\n",
    "        dna = dna.replace(dna[One_slice], '')\n",
    "        L=len(OrgDna)-len(dna)\n",
    "        if i == len(intron):\n",
    "            break\n",
    "    return(dna)\n",
    "\n",
    "\n",
    "def splice_rna(dna, intron_list):\n",
    "    protein = ''\n",
    "    \n",
    "    i = 0\n",
    "    intron=[]\n",
    "    # return the postion of the introns\n",
    "    dna = intron_cut(dna, intron_list)\n",
    "    # Convert the string from DNA to RNA.\n",
    "    rna = dna2rna(dna)\n",
    "    \n",
    "    # Convert the RNA string to its corresponding protein expression string.\n",
    "    protein = rna2codon(rna)\n",
    "    \n",
    "    # Return the resulting string.\n",
    "    return (protein)\n",
    "\n",
    "#Meena\n",
    "\n",
    "# s(dna) function\n",
    "# The purpose of this function is to count the number of nucleotides in a given strand and return a dictionary showing the counts for each nucleotide. \n",
    "dna =''\n",
    "def s(dna):\n",
    "    countA = dna.count('A')\n",
    "    countC = dna.count('C')\n",
    "    countG = dna.count('G')\n",
    "    countT = dna.count('T')\n",
    "    s = {\n",
    "        'A': countA,\n",
    "        'C': countC, \n",
    "        'G': countG, \n",
    "        'T': countT\n",
    "    }\n",
    "    return(s)\n",
    "\n",
    "\n",
    "# This function will return the hamming distance between 2 strands of nucleotides\n",
    "dna1 = \"\"\n",
    "dna2 = \"\"\n",
    "def hamming_dist(dna1,dna2):\n",
    "    count = 0\n",
    "    x = len(dna1)\n",
    "    for nucleotide in range(0,x):\n",
    "        if dna1[nucleotide] != dna2[nucleotide]:\n",
    "            count += 1\n",
    "    return count\n",
    "\n",
    "\n",
    "# This function will calculate the total number of rabbit pairs present after n months given that \n",
    "# each pair produces k offspring\n",
    "def fibonacci_rabbits(n,k):\n",
    "    if n == 1:\n",
    "        return (1)\n",
    "    elif n == 2:\n",
    "        return (1)\n",
    "    elif n >= 3:\n",
    "        return (fibonacci_rabbits(n-1,k) + k*fibonacci_rabbits(n-2,k))\n",
    "    else:\n",
    "        return (none)\n",
    "    \n",
    "#Zylle\n",
    "\n",
    "def count_dom_phenotype(genotypes):\n",
    "    return (genotypes[0]*2) + (genotypes[1]*2) + (genotypes[2]*2) + (genotypes[3]*1.5) + (genotypes[4]*1) + (genotypes[5]*0)\n",
    "\n",
    "# this function is supposed to turn dna into rna strands\n",
    "# specifically turns \"T\" into \"U\" in the dna strands\n",
    "\n",
    "def dna2rna(dna):\n",
    "    rna = \"\"\n",
    "    for symbol in dna:\n",
    "        if symbol == \"A\":\n",
    "            rna = rna + \"A\"\n",
    "        elif symbol == \"T\":\n",
    "            rna = rna + \"U\"\n",
    "        elif symbol == \"C\":\n",
    "            rna = rna + \"C\"\n",
    "        elif symbol == \"G\":\n",
    "            rna = rna + \"G\"\n",
    "    return rna\n",
    "\n",
    "\n",
    "#this function is supposed to accept a list of DNA strings & return the index\n",
    "#of which string has the highest \"GC\" content, along with its percentage of\n",
    "#\"C\" or \"G\" in the specified string.\n",
    "\n",
    "\n",
    "def gc_content(dna_list):\n",
    "    counts = []\n",
    "    for index in range(len(dna_list)):\n",
    "        count_G = dna_list[index].count(\"G\")\n",
    "        count_C = dna_list[index].count(\"C\")\n",
    "        count = count_C + count_G\n",
    "        counts.append((index,100*count/len(dna_list[index])))\n",
    "    curr_max = (0,0)\n",
    "    for elem in counts:\n",
    "        if (elem[1] > curr_max[1]):\n",
    "            curr_max = elem\n",
    "    return curr_max"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "28a8c1a1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
